### PR TITLE
StatefulSet fsGroup config option to allow non-root spilo

### PR DESCRIPTION
**Pull request by aaroniscode**
xref #530 

Interested in your thoughts around this approach and whether defaulting fsGroup to 0, which doesn't provide any securityContext to the pods in the StatefulSet is the preferred approach.

### CODE REVIEW

**Comment by erthalion** [comment]
Thanks for the PR! It looks reasonable, I'll test in in the next few days and by the results of this test we can hopefully merge it.

**Comment by erthalion** [comment]
While testing it with `spiloFSGroup = 0` for some reason I've got a strange error from patroni (when it tries to list pods), that I can't see without the change:

```
pods is forbidden: User "system:serviceaccount:default:postgres-operator" cannot list resource "pods"
```

While I'm checking, if I'm doing something wrong, @aaroniscode have you encountered such a problem?

**Comment by aaroniscode** [comment]
@erthalion A quick review of the code and I can't see how a change to the securityContext of Spilo would impact a permissions capability. I would think that would be a RBAC rule. I never encountered this in my testing, but I'll take another look today.

Is it possible that the service account is just lacking the permissions due to some other configuration or issue?

**Comment by erthalion** [comment]
> Is it possible that the service account is just lacking the permissions due to some other configuration or issue?

Yes, I guess something like that, otherwise it's hard to explain. I'll try to figure this out, just to make sure that it has nothing to do with the PR.

**Comment by aaroniscode** [comment]
@erthalion Let me know if you aren't able to resolve the issue and it's looking like it's related to the PR. At the moment I can't see how it would be related, but maybe I'm missing something. I can try to track it down on my side if you aren't having any luck.

**Comment by erthalion** [comment]
Yes, it's not related - looks like it was a minor hiccup with service accounts, after I've rebased the code on the latest master everything seems fine. I'll do few more tests with/without fs group, and then approve the pr.

**Comment by aaroniscode** [comment]
Great news. Thanks for the update!

**Comment by erthalion** [comment]
Thanks @aaroniscode, at the end it looks good, I only have a few questions:

* I couldn't find it in the K8S docs or the source code, but from my tests it looks like it's not possible to change `FSGroup` after creation, is it correct?
* This options depends on a spilo image, but one can change it in both operator and database configuration, while in this PR it's possible only on the operator level (so, globally). Is it worth it to allow change it on database level too?
* I would like to ask you to add this option to the configuration CRD and Help chart as well (as described here [1]), since operator can be configured via all of them. Could you please do this? 

[1]: https://github.com/zalando/postgres-operator/blob/master/docs/developer.md#introduce-additional-configuration-parameters

**Comment by aaroniscode** [comment]
> * I couldn't find it in the K8S docs or in the source code, but from my tests it looks like it's not possible to change `FSGroup` after creation, is it correct? (If yes, it's probably worth to document this)

@erthalion I just tested and confirmed FSGroup cannot be changed after creation.

> * This options depends on a spilo image, but one can change it in both operator and database configuration, while in this PR it's possible only on the operator level (so, globally). Is it worth it to allow change it on database level too?

I wasn't aware you could change it at the database configuration. I think it would be worth it to allow it to change on the database level too. I'll see how I can add that to the PR.

> * I would like to ask you to add this option to the configuration CRD and Help chart as well (as described here [1](https://github.com/zalando/postgres-operator/blob/master/docs/developer.md#introduce-additional-configuration-parameters)), since operator can be configured via all of them. Could you please do this?

Not a problem. I'll add that as well.

I'll work on this early next week. Hope the timing works for you.

**Comment by aaroniscode** [comment]
Quick update -- I believe I've got the code updated to reflect the changes requested. I going to test some more tomorrow morning then push the changes by mid-day EDT (New York time).

**Comment by aaroniscode** [comment]
Found an issue with the original implementation of using a `int64`. If the Operator used a non-root FSGroup like 103 which is the `postgres` group, there was no way for the Postgres CRD to use a root FSGroup of 0 because 0 is also the default value.

So switched it over to a `*int64` and that way `nil` represents the default.

I'm running up on a deadline as I am on vacation starting tomorrow so while I was able to do some quick testing, I didn't get to test as much as I wanted. But I think this is working as intended. Let me know if I missed anything or if you think this needs any design changes. Thanks!

**Comment by aaroniscode** [comment]
I'm not sure why the build is failing. I confirmed the build works on my branch and I'm using go 1.12.4.

If there is something wrong on my side, just point me in the right direction and I'll fix it ASAP.

**Comment by erthalion** [comment]
> I'm not sure why the build is failing. I confirmed the build works on my branch and I'm using go 1.12.4.

I'm confused myself, since for me locally everything is fine too. Investigating right now, but to manage the expectations - probably we will not be able to merge it today. Is it ok from your side?

**Comment by aaroniscode** [comment]
> I'm confused myself, since for me locally everything is fine too. Investigating right now, but to manage the expectations - probably we will not be able to merge it today. Is it ok from your side?

No worries. I'll be available next week if you need any changes once you have a chance to review. Thanks for considering this PR.

**Comment by erthalion** [comment]
Oh, of course - the tests are failing for reason. @aaroniscode there is another `generatePodTemplate` invocation in the master right now (came with logical backup feature) that should be updated with `spiloFSGroup`. And when Travis clones the PR it clones the merge commit (where this function call is present) and fails to build it. The solution would be just rebase and fix the error.

**Comment by aaroniscode** [comment]
> Oh, of course - the tests are failing for reason. @aaroniscode there is another `generatePodTemplate` invocation in the master right now (came with logical backup feature) that should be updated with `spiloFSGroup`. And when Travis clones the PR it clones the merge commit (where this function call is present) and fails to build it. The solution would be just rebase and fix the error.

Well, that makes sense! So, I read through the logical backup PR. I think we have a number of options of how we want to proceed.

1. The logical backup CronJob is created without specifying the FSGroup.
2. The logical backup CronJob assumes the Spilo pod's FSGroup.
3. The logical backup CronJob gets it's own FSGroup because it can also override the docker image.

I implemented the quick fix using Option #1. I don't think Option #2 or #3 are needed as I don't believe the pod is assigned a Persistent Volume.

Just let me know if I'm missing something as I quickly skimmed the logical backup PR.

**Comment by erthalion** [comment]
Yes, I believe option 1 is what we need.

**Comment by aaroniscode** [comment]
@erthalion just wanted to check in and see if you need anything else on this PR?

I have already implemented Option 1 that I described above. One of the checks is failing because coverage decreased -- I think this is because of the generators creating a new function.

**Comment by erthalion** [comment]
> @erthalion just wanted to check in and see if you need anything else on this PR?

I'm testing the PR right now. The only thing I wanted to ask is that looks like the documentation for this:

> I couldn't find it in the K8S docs or in the source code, but from my tests it looks like it's not possible to change FSGroup after creation, is it correct? (If yes, it's probably worth to document this)

is still missing. Would be nice to mention in a few words that this can't be changed on the fly.

File: manifests/postgresql-operator-default-configuration.yaml [hunk 1]
```diff
@@ -23,6 +23,7 @@ configuration:
     secret_name_template: "{username}.{cluster}.credentials.{tprkind}.{tprgroup}"
     oauth_token_secret_name: postgresql-operator
     pod_role_label: spilo-role
+    # spilo_fsgroup: "103"
```
Comment (erthalion) [inline manifests/postgresql-operator-default-configuration.yaml]: I think this should be without quotes, otherwise `readUint64` would complain.

Comment (aaroniscode) [reply manifests/postgresql-operator-default-configuration.yaml]: Removed the quotes!

**Comment by aaroniscode** [comment]
There were some conflicts with master so I had to rebase. I believe I addressed all the recent items in the last few days.

> > I couldn't find it in the K8S docs or in the source code, but from my tests it looks like it's not possible to change FSGroup after creation, is it correct? (If yes, it's probably worth to document this)
> 
> is still missing. Would be nice to mention in a few words that this can't be changed on the fly.

Added to the docs!

Thanks for testing the PR and reviewing.

**Comment by erthalion** [comment]
> There were some conflicts with master so I had to rebase. I believe I addressed all the recent items in the last few days.

Great, thank you! I hope I can make a final review tomorrow.

**Comment by erthalion** [comment]
:+1:

**Comment by sdudoladov** [comment]
:+1:

**Comment by sdudoladov** [comment]
@aaroniscode thanks for the contribution !

**Pull request merged**